project-icp/bee-pollinator-app
# User survey form

## Overview

After a user logs in the for the first time to the beekeepers app, they need to fill out a user survey that details some basic information. The user shouldn't be able to use the app or exit without filling out this survey, though the fields are mostly optional.

The only field that isn't optional (from the UI's perspective) is `year_began`. I did this only because the django model expects a positive int. In the UI, the text input can't be given a default value, even null, `''`' or `NaN`. It was easier to just require this simple piece of info.

Connects #383

### Demo

<img width="636" alt="screen shot 2018-12-28 at 4 04 21 pm" src="https://user-images.githubusercontent.com/10568752/50528085-463b5880-0aba-11e9-9efd-39030791ca01.png">


### Notes

Forms are crazy!! Took a long time to get just right enough... for all who do a survey form, we should pair because i definitely learned a thing or two here that would save time in the future (but not worth writing down perse)

Most importantly though was loosening up the Django model to be form-ready: basically, fields accepting blanks in most cases. See the commit message or https://stackoverflow.com/questions/8609192/differentiate-null-true-blank-true-in-django

If the user survey fails, the error handling is a generic message -- which isn't helpful. Unlike logging in, however, the error messages that I've seen come back were not very helpful in context or tough to parse/not worth it.

## Testing Instructions
`./scripts/manage migrate`
`./scripts/beekeepers.sh start`
Assuming you have a user, log in to beekeepers. 
- Logging in when you don't have the user survey yet should prompt the user survey modal
The survey should submit just fine without entering more than the prompted required field(s).

- Logging into Bees 1 should work normally
- Logging in when you already have a user survey shouldn't prompt the user survey modal

FYI the new endpoint is at `/beekeepers/user_survey/`
